<a href="https://colab.research.google.com/github/FunmiSomoye/School-Projects/blob/master/Feature_Engineering_MLP_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# HLT Assignment

## HLT_Group-C

#### Raphael Ozioko
#### Tobechukwu Asem
#### Oluwafunmilayo Somoye

In addition to practising the process of loading data, regularizing it and using it to train a model, the goal of this assignment is to investigat the effect of changing model parameters on the performance.

## Step 1: loading the data

We'll be going a bit further with the robot collision dataset. This time, instead of looking at just the first file, we'll look at all five different tasks combined (lp1.data to lp5.data). Prepare two different arrays, X1 and X2, as follows:
- Each element in X1 is the immediate reading of the force and torque values after an event, [f1, f2, f3, t1, t2, t3]. The first element should be [1, 1, 63, 3, 1, 0]
- Each element in X2 contains 18 values in total - the first, fifth and tenth sets of sensor readings after an event. The first element should be [-1, -1, 63, -3, -1, 0, -1, -1, 63, -3, -1, 0, -1, -1, 61, -3, 0, 0].

y should contain the corresponding classes, represented as integers according the the provided dictionary

In [ ]:
files = ['robot_execution_failure/lp1.data', 
         'robot_execution_failure/lp2.data',
         'robot_execution_failure/lp3.data', 
         'robot_execution_failure/lp4.data', 
         'robot_execution_failure/lp5.data'
        ]
classes = {'normal':0, 'collision':1, 'obstruction':2, 'fr_collision':3} # Encoding the classes as integers

In [ ]:
#import needed libraries here
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import math

In [ ]:
# Your answer

raw_data = []

for filename in files:
    with open(filename) as f:
        for line in [item for item in f.readlines()]:# Read the file line by line into a list
            raw_data.append(line)

In [ ]:
raw_data[:5]#Read first 5 lines

['normal\n',
 '\t-1\t-1\t63\t-3\t-1\t0\n',
 '\t0\t0\t62\t-3\t-1\t0\n',
 '\t-1\t-1\t61\t-3\t0\t0\n',
 '\t-1\t-1\t63\t-2\t-1\t0\n']

In [ ]:
len(raw_data)

8326

##### Create X1

In [ ]:
X1 = [] # inputs
y = [] # true values

# Here, we iterate over the lines of the file. If a line matches one of our classes, we split the next line 
# to get the six readings and use those as our features. 
for i in range(len(raw_data) - 1):
    data = raw_data[i].strip() # .strip() removes the line endings \n
    if data in classes.keys(): # If the line matches one of our classes (for eg, 'normal')
        features = [int(x) for x in raw_data[i+1].strip().split('\t')] # Split the next line to get our features
        X1.append(features)
        y.append(classes[data]) # And record which class this set of features belongs to

In [ ]:
len(X1)

269

In [ ]:
X1[0]

[-1, -1, 63, -3, -1, 0]

In [ ]:
y

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 0,
 2,
 0,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 3,
 2,
 3,
 3,
 3,
 2,
 3,
 2,
 2,
 2,
 3,
 2,
 2,
 2,
 2,
 0,
 3,
 3,
 2,
 3,
 2,
 2,
 3,
 2,
 3,
 2,
 2,
 3,
 2,
 3,
 3,
 2,
 3,
 2,
 2,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 2,
 2,
 2,
 2,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 1,
 2,
 1,
 2,
 1,
 2,
 2,
 1,
 2,
 1,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,


##### Create X2

In [ ]:
X2 = [] # inputs
y = [] # true values

# Here, we iterate over the lines of the file. If a line matches one of our classes, we split the next line 

for i in range(len(raw_data) - 1):
    data2 = raw_data[i].strip() # .strip() removes the line endings 
    if data2 in classes.keys(): # If the line in the data matches one of our classes (for eg, 'normal')
        features = [int(x) for x in raw_data[i+1].strip().split('\t') + raw_data[i+5].strip().split('\t') + raw_data[i+10].strip().split('\t')] # Split the next line to get our features
        X2.append(features)
        y.append(classes[data2]) # And record which class this set of features belongs to

In [ ]:
X2[0]

[-1, -1, 63, -3, -1, 0, -1, -1, 63, -3, -1, 0, -1, -1, 61, -3, 0, 0]

## Step 2: establishing a baseline

Using techniques covered in this unit, split X1 and y into separate training and testing sets. Use the training set to train a neural network (MLPClassifier) using default parameters but with hidden_layer_sizes=(20, 20, 20). Use the test data you held back to score the model you have created. How well does it perform? Print out the score and confusion matrix. For more accuracy, run through these steps 10 times and find the average score - bonus points for running more times and getting a standard deviation!

In [ ]:
# Your answer
#creaate coppies of X1 for multiple iterations
X1_1 = np.copy(X1)
X1_2 = np.copy(X1)
X1_3 = np.copy(X1)

##### 1

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X1_1, y)  #split into training and test sets

In [ ]:
len(X_train) #see size of X_train

201

In [ ]:
# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


StandardScaler(copy=True, with_mean=True, with_std=True)

In [ ]:
# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [ ]:
#import the MLP Classifier
from sklearn.neural_network import MLPClassifier

In [ ]:
# Creating the neural network
mlp1 = MLPClassifier(hidden_layer_sizes=(20,20,20))
mlp1.fit(X_train,y_train)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(20, 20, 20), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [ ]:
predict1 = mlp1.predict(X_test) #predict

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix #import functions for score calculations


In [ ]:
# The score - pretty good! ~96% accuracy (will vary depending on the random weights the model started with)
print(mlp1.score(X_test, y_test))

print(confusion_matrix(y_test,predict1))
print(classification_report(y_test,predict1))

0.6470588235294118
[[26  0  0  0]
 [ 8 17  1  0]
 [ 6  6  1  0]
 [ 1  2  0  0]]
              precision    recall  f1-score   support

           0       0.63      1.00      0.78        26
           1       0.68      0.65      0.67        26
           2       0.50      0.08      0.13        13
           3       0.00      0.00      0.00         3

   micro avg       0.65      0.65      0.65        68
   macro avg       0.45      0.43      0.39        68
weighted avg       0.60      0.65      0.58        68



C:\Users\acer\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


If you're getting a convergence warning, you could try having the model train over more iterations - change max_iter = 1000 or 10,000. Does this improve the average score?

In [ ]:
#using max_iter of 1000

##### 2

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X1_2, y) #split into training and test sets

In [ ]:
# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [ ]:
# Creating the neural network
mlp2 = MLPClassifier(hidden_layer_sizes=(20,20,20), max_iter=1000)
mlp2.fit(X_train,y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(20, 20, 20), learning_rate='constant',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [ ]:
predict2 = mlp2.predict(X_test) #predict

In [ ]:
# The score - pretty good! ~96% accuracy (will vary depending on the random weights the model started with)
print(mlp2.score(X_test, y_test))
# The confusion matrix gives a better picture of what errors are slipping in

print(confusion_matrix(y_test,predict2))
print(classification_report(y_test,predict2))

0.6617647058823529
[[29  0  0  0]
 [10 13  2  0]
 [ 2  7  3  0]
 [ 2  0  0  0]]
              precision    recall  f1-score   support

           0       0.67      1.00      0.81        29
           1       0.65      0.52      0.58        25
           2       0.60      0.25      0.35        12
           3       0.00      0.00      0.00         2

   micro avg       0.66      0.66      0.66        68
   macro avg       0.48      0.44      0.43        68
weighted avg       0.63      0.66      0.62        68



C:\Users\acer\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [ ]:
#using max_iter of 10000

##### 3

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X1_3, y) #split into training and test sets

In [ ]:
len(X_train)#see size of X_train

201

In [ ]:
# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [ ]:
# Creating the neural network
mlp3 = MLPClassifier(hidden_layer_sizes=(20,20,20), max_iter=10000)
mlp3.fit(X_train,y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(20, 20, 20), learning_rate='constant',
       learning_rate_init=0.001, max_iter=10000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [ ]:
predict3 = mlp3.predict(X_test) #predict

In [ ]:
# The score - pretty good! ~96% accuracy (will vary depending on the random weights the model started with)
print(mlp3.score(X_test, y_test))
# The confusion matrix gives a better picture of what errors are slipping in
print(confusion_matrix(y_test,predict3))
print(classification_report(y_test,predict3))

0.6470588235294118
[[25  3  0  0]
 [ 5 16  2  0]
 [ 5  4  3  0]
 [ 4  0  1  0]]
              precision    recall  f1-score   support

           0       0.64      0.89      0.75        28
           1       0.70      0.70      0.70        23
           2       0.50      0.25      0.33        12
           3       0.00      0.00      0.00         5

   micro avg       0.65      0.65      0.65        68
   macro avg       0.46      0.46      0.44        68
weighted avg       0.59      0.65      0.60        68



C:\Users\acer\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [ ]:
#max_iter 1000 has the highest mlp score of 0.66 so far
#continue testing with max_iter of 1000
#conduct 20 tests

##### Multiple Iterations

In [ ]:
# Creating the neural network
#using a for loop to run through 20 times

X_train, X_test, y_train, y_test = train_test_split(X1, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)


mlpscores = [] #store scores here

for counter in range(1,21): #iterate 20 times
    mlp = MLPClassifier(hidden_layer_sizes=(20,20,20), max_iter=1000)
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

The mlp Score is  0.7058823529411765
The confusion matrix is  [[24  0  0  0]
 [ 5 19  1  0]
 [ 2  7  5  0]
 [ 5  0  0  0]]

The mlp Score is  0.6470588235294118
The confusion matrix is  [[24  0  0  0]
 [10 14  1  0]
 [ 7  1  6  0]
 [ 5  0  0  0]]

The mlp Score is  0.5882352941176471
The confusion matrix is  [[24  0  0  0]
 [13 11  1  0]
 [ 9  0  5  0]
 [ 5  0  0  0]]

The mlp Score is  0.6470588235294118
The confusion matrix is  [[24  0  0  0]
 [10 14  1  0]
 [ 6  2  6  0]
 [ 5  0  0  0]]

The mlp Score is  0.6323529411764706
The confusion matrix is  [[24  0  0  0]
 [10 13  2  0]
 [ 7  1  6  0]
 [ 5  0  0  0]]

The mlp Score is  0.6176470588235294
The confusion matrix is  [[20  4  0  0]
 [ 7 17  1  0]
 [ 5  4  5  0]
 [ 5  0  0  0]]

The mlp Score is  0.5588235294117647
The confusion matrix is  [[24  0  0  0]
 [13  9  3  0]
 [ 7  2  5  0]
 [ 5  0  0  0]]

The mlp Score is  0.6323529411764706
The confusion matrix is  [[24  0  0  0]
 [ 6 19  0  0]
 [ 5  9  0  0]
 [ 5  0  0  0]]

The mlp 

In [ ]:
len(mlpscores) #confirm that the loop ran 20 times

20

In [ ]:
mlpscores

[0.7058823529411765,
 0.6470588235294118,
 0.5882352941176471,
 0.6470588235294118,
 0.6323529411764706,
 0.6176470588235294,
 0.5588235294117647,
 0.6323529411764706,
 0.6470588235294118,
 0.6764705882352942,
 0.6617647058823529,
 0.6029411764705882,
 0.6029411764705882,
 0.6617647058823529,
 0.6176470588235294,
 0.5735294117647058,
 0.6617647058823529,
 0.5294117647058824,
 0.6617647058823529,
 0.5882352941176471]

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) #print scores and round up numbers to 2 decimal places

The average mlp score is = 0.63


In [ ]:
#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The std of the mlp scores is = 0.04


An average mlp score of 0.63 above does not depict an excellent score. Based on this score, this isn't a good model. 

## Step 3: adding more inputs

Use X2 in place of X1 - does the score increase or decrease? Was this what you expected? How many samples are there in our training set?

In [ ]:
# Your answer
#creaate a copy of X2 to save original list
X2_1 = np.copy(X2)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X2_1, y) #split into training and test sets

In [ ]:
len(X_train)#see size of X_train

201

There are 201 samples in the training set

In [ ]:
# Prepare to scale the inputs
scaler2 = StandardScaler()
scaler2.fit(X_train)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


StandardScaler(copy=True, with_mean=True, with_std=True)

In [ ]:
# Scaling the inputs 
X_train = scaler2.transform(X_train)
X_test = scaler2.transform(X_test)

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [ ]:
# Creating the neural network
#using a for loop to run through 20 times

mlpscores_X2 = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    mlp_X2 = MLPClassifier(hidden_layer_sizes=(20,20,20), max_iter=1000)
    mlp_X2.fit(X_train,y_train)
    predictions = mlp_X2.predict(X_test)
    
    scores_X2 = mlp_X2.score(X_test, y_test)
    mlpscores_X2 += [scores_X2]
    
    print('The mlp Score is ',scores_X2)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

The mlp Score is  0.7794117647058824
The confusion matrix is  [[26  0  0  0]
 [ 6 20  0  0]
 [ 1  3  7  0]
 [ 3  2  0  0]]

The mlp Score is  0.7352941176470589
The confusion matrix is  [[26  0  0  0]
 [ 6 20  0  0]
 [ 1  6  4  0]
 [ 2  3  0  0]]

The mlp Score is  0.6617647058823529
The confusion matrix is  [[26  0  0  0]
 [ 8 14  4  0]
 [ 3  3  5  0]
 [ 3  2  0  0]]

The mlp Score is  0.6764705882352942
The confusion matrix is  [[26  0  0  0]
 [10 14  2  0]
 [ 4  1  6  0]
 [ 4  1  0  0]]

The mlp Score is  0.7352941176470589
The confusion matrix is  [[26  0  0  0]
 [ 7 19  0  0]
 [ 3  3  5  0]
 [ 3  2  0  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[26  0  0  0]
 [ 4 21  1  0]
 [ 1  4  6  0]
 [ 2  3  0  0]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[25  1  0  0]
 [ 5 20  1  0]
 [ 0  4  7  0]
 [ 2  3  0  0]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[25  1  0  0]
 [ 3 22  1  0]
 [ 0  4  7  0]
 [ 1  4  0  0]]

The mlp 

In [ ]:
mlpscores_X2

[0.7794117647058824,
 0.7352941176470589,
 0.6617647058823529,
 0.6764705882352942,
 0.7352941176470589,
 0.7794117647058824,
 0.7647058823529411,
 0.7941176470588235,
 0.7647058823529411,
 0.7647058823529411,
 0.7058823529411765,
 0.7205882352941176,
 0.8088235294117647,
 0.7794117647058824,
 0.7352941176470589,
 0.7794117647058824,
 0.7205882352941176,
 0.7205882352941176,
 0.7941176470588235,
 0.7794117647058824]

In [ ]:
#compute average of mlp scores obtained
average_X2 = np.mean(mlpscores_X2) #calculate average
print ("The average mlp score is =", round(average_X2, 2)) 

#compute std of mlp scores obtained
std_X2 = np.std(mlpscores_X2)
print ("The std of the mlp scores is =", round(std_X2, 2)) 

The average mlp score is = 0.75
The std of the mlp scores is = 0.04


The performance of X2 is better as a majority of the scores fell between 0.70 to 0.77. This model is not optimum, but performs better than with X1 input parameters.

## Step 4: feature engineering

Back to X1 as our input. Add an extra feature to each item in the array to represent the total force $f_t$. Assume:

$f_t^2 = f_1^2 + f_2^2 + f_3^2$

Your first input should now look like this:
X1[0] = [-1, -1, 61, -3, 0, 0, 61.0163912403872]

Repeat the steps from step 2. *Has this extra feature improved model performance?*

In [ ]:
X1

[[-1, -1, 63, -3, -1, 0],
 [-1, -1, 63, -2, -1, 0],
 [-1, 0, 57, -5, -3, 0],
 [0, -1, 59, -2, -1, -1],
 [0, -2, 65, -4, -2, 0],
 [-1, 2, 57, -10, -3, 0],
 [-3, 1, 53, -10, -4, 0],
 [2, 0, 69, -10, 2, 1],
 [-1, 1, 58, -11, -4, -1],
 [-1, -1, 63, -3, -1, 0],
 [-1, -1, 63, -3, 0, 0],
 [1, -3, 62, -1, 2, 0],
 [0, -1, 69, -4, 2, -1],
 [1, -2, 57, -4, 0, -1],
 [-1, 1, 67, -9, -4, -1],
 [0, 2, 67, -11, -3, 0],
 [-1, 1, 58, -8, -4, 1],
 [1, 0, 60, -8, 1, 1],
 [24, 53, 79, -200, 109, 2],
 [185, 105, 118, -180, 374, 12],
 [-21, -50, 46, -180, 266, 98],
 [112, -54, 85, -77, 284, -3],
 [95, -4, 38, 125, 111, -4],
 [-127, 83, 69, -71, -148, -53],
 [-72, -24, 84, 120, -196, -15],
 [-133, -34, 18, 21, 131, -41],
 [-10, -2, 27, -17, 8, 3],
 [5, 23, 157, -186, 179, 2],
 [-14, -10, 52, 15, -77, 0],
 [8, -17, 50, 16, -1, -2],
 [3, -33, 32, 41, 2, -1],
 [2, -3, 70, -3, -1, 0],
 [4, -14, 59, 82, -85, 13],
 [71, 51, 5, 686, 494, -137],
 [-93, 76, 80, -123, -151, -1],
 [-8, 20, -5, -26, -10, -1],
 [11, -3, -

In [ ]:
#make another copy of the original data
X1_4 = np.copy(X1)

In [ ]:
#x2 = np.sum(X1_4, axis = 1, keepdims = True) #creates a multidimensional array

In [ ]:
x = np.sum(X1_4[:, :3] ,axis=1) #sum the columns (axis=1)
x.T #tranpose from horizontal to vertical



array([   61,    61,    56,    58,    63,    58,    51,    71,    58,
          61,    61,    60,    68,    56,    67,    69,    58,    61,
         156,   408,   -25,   143,   129,    25,   -12,  -149,    15,
         185,    28,    41,     2,    69,    49,   127,    63,     7,
        -293,  -344,  -405,  -429,    52,  -883,    63, -1047,    58,
        -267,  -583,   -14,  -715,  -790,    19,  -205,  -104,  -429,
          53,    30,     1,  -273,    71,   -81,  -393, -1463,    50,
         -69,  -400,    59,    41,    57,    40,     0,  -220,    45,
          12,  -301,    37,   -36,    48,   -23,  -412,    41,    13,
          43,   -50,  -247,    60,    22,  -229,    49,    78,    84,
          77,    74,    76,    71,    76,    77,    75,    82,    76,
          83,    68,    75,    72,    74,    71,    77,    64,    74,
          20,    21,    20,    20,    20,    18,    19,    18,    22,
          16,     5,    19,    18,    19,    16,    19,    18,    18,
           4,     4,

In [ ]:
new_X1 = np.insert(X1_4,#array
                   6,#position to input new elements
                   (x),#we could have listed out numbers, but they are too many. so just use x (which contains all the numbers) 
                   axis=1)#add it as a new column across the matrix

In [ ]:
new_X1 [0]

array([-1, -1, 63, -3, -1,  0, 61])

In [ ]:
# Creating the neural network
#using a for loop to run through 20 times

X_train, X_test, y_train, y_test = train_test_split(new_X1, y)

# Prepare to scale the inputs
scaler3 = StandardScaler()
scaler3.fit(X_train)

# Scaling the inputs 
X_train = scaler3.transform(X_train)
X_test = scaler3.transform(X_test)


mlpscores_newX1 = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    mlp_newX1 = MLPClassifier(hidden_layer_sizes=(20,20,20), max_iter=1000)
    mlp_newX1.fit(X_train,y_train)
    predictions = mlp_newX1.predict(X_test)
    
    scores_newX1 = mlp_newX1.score(X_test, y_test)
    mlpscores_newX1 += [scores_newX1]
    
    print('The mlp Score is ',scores_newX1)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.6764705882352942
The confusion matrix is  [[30  0  0  0]
 [ 8 12  0  0]
 [ 4  4  4  0]
 [ 5  1  0  0]]

The mlp Score is  0.7352941176470589
The confusion matrix is  [[30  0  0  0]
 [ 3 16  1  0]
 [ 1  7  4  0]
 [ 4  2  0  0]]

The mlp Score is  0.75
The confusion matrix is  [[29  1  0  0]
 [ 1 18  1  0]
 [ 3  5  4  0]
 [ 3  3  0  0]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[30  0  0  0]
 [ 5 14  1  0]
 [ 4  3  5  0]
 [ 5  1  0  0]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[30  0  0  0]
 [ 3 17  0  0]
 [ 3  4  5  0]
 [ 3  3  0  0]]

The mlp Score is  0.6764705882352942
The confusion matrix is  [[23  7  0  0]
 [ 1 19  0  0]
 [ 0  8  4  0]
 [ 2  4  0  0]]

The mlp Score is  0.6617647058823529
The confusion matrix is  [[30  0  0  0]
 [ 8 11  1  0]
 [ 5  3  4  0]
 [ 5  1  0  0]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[29  1  0  0]
 [ 0 19  1  0]
 [ 1  5  6  0]
 [ 5  1  0  0]]

The mlp Score is  0.66

In [ ]:
mlpscores_newX1

[0.6764705882352942,
 0.7352941176470589,
 0.75,
 0.7205882352941176,
 0.7647058823529411,
 0.6764705882352942,
 0.6617647058823529,
 0.7941176470588235,
 0.6617647058823529,
 0.7205882352941176,
 0.7352941176470589,
 0.7647058823529411,
 0.6764705882352942,
 0.6617647058823529,
 0.4411764705882353,
 0.7205882352941176,
 0.7058823529411765,
 0.6764705882352942,
 0.7058823529411765,
 0.7794117647058824]

In [ ]:
#compute average of mlp scores obtained
average_newX1 = np.mean(mlpscores_newX1) #calculate average
print ("The average mlp score is =", round(average_newX1, 2)) 

#compute std of mlp scores obtained
std_newX1 = np.std(mlpscores_newX1)
print ("The std of the mlp scores is =", round(std_newX1, 2)) 

The average mlp score is = 0.7
The std of the mlp scores is = 0.07


The above shows that this model did better than the first model for X1. Nevertheless, it still is not good enough as the scores are not in the 90% range.
The Standard deviation here is worse than for the first X1 model.

## Step 5: exploring

Create at least 3 more models, adding features or changing the number and size of the hidden layers. Print out the average score for your best model. Comment on what you've found.

In [ ]:
# Your answer

##### Model 1 extra

In [ ]:
# CHANGING THE SIZES OF THE HIDDEN LAYERS, using X1
# Creating the neural network
#using a for loop to run through 20 times

X_train, X_test, y_train, y_test = train_test_split(X1, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    mlp = MLPClassifier(hidden_layer_sizes=(30,30,30), max_iter=1000)
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

The mlp Score is  0.8088235294117647
The confusion matrix is  [[31  1  0  0]
 [ 1 12  4  2]
 [ 1  2 12  0]
 [ 1  0  1  0]]

The mlp Score is  0.8235294117647058
The confusion matrix is  [[31  0  0  1]
 [ 1 13  3  2]
 [ 1  2 12  0]
 [ 1  0  1  0]]

The mlp Score is  0.8235294117647058
The confusion matrix is  [[32  0  0  0]
 [ 2 12  4  1]
 [ 1  2 12  0]
 [ 1  0  1  0]]

The mlp Score is  0.8088235294117647
The confusion matrix is  [[32  0  0  0]
 [ 1 10  7  1]
 [ 1  1 13  0]
 [ 1  0  1  0]]

The mlp Score is  0.8382352941176471
The confusion matrix is  [[32  0  0  0]
 [ 1 12  5  1]
 [ 1  1 13  0]
 [ 1  0  1  0]]

The mlp Score is  0.8088235294117647
The confusion matrix is  [[32  0  0  0]
 [ 3  9  6  1]
 [ 1  0 14  0]
 [ 1  0  1  0]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[30  0  0  2]
 [ 0 14  2  3]
 [ 1  6  8  0]
 [ 1  0  1  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[32  0  0  0]
 [ 5  8  5  1]
 [ 1  0 13  1]
 [ 1  0  1  0]]

The mlp 

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.8
The std of the mlp scores is = 0.05


In [ ]:
Increasing the size of the hidden layers deifinitely increased the model performance by a lot!!!

##### Model 2 extra

In [ ]:
# CHANGING THE SIZE OF THE HIDDEN LAYERS, using X2
# Creating the neural network
#using a for loop to run through 20 times
X2_2 = np.copy(X2)

X_train, X_test, y_train, y_test = train_test_split(X2_2, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    mlp = MLPClassifier(hidden_layer_sizes=(30,30,30), max_iter=1000)
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.7794117647058824
The confusion matrix is  [[28  0  0  0]
 [ 2 15  1  0]
 [ 0  7 10  0]
 [ 1  3  1  0]]

The mlp Score is  0.8235294117647058
The confusion matrix is  [[27  1  0  0]
 [ 0 16  2  0]
 [ 0  4 13  0]
 [ 2  1  2  0]]

The mlp Score is  0.8235294117647058
The confusion matrix is  [[28  0  0  0]
 [ 1 16  1  0]
 [ 1  4 12  0]
 [ 1  3  1  0]]

The mlp Score is  0.75
The confusion matrix is  [[28  0  0  0]
 [ 2 14  2  0]
 [ 3  5  9  0]
 [ 2  2  1  0]]

The mlp Score is  0.8235294117647058
The confusion matrix is  [[25  3  0  0]
 [ 0 18  0  0]
 [ 0  4 13  0]
 [ 1  3  1  0]]

The mlp Score is  0.6764705882352942
The confusion matrix is  [[28  0  0  0]
 [ 9  9  0  0]
 [ 3  5  9  0]
 [ 3  1  1  0]]

The mlp Score is  0.8529411764705882
The confusion matrix is  [[28  0  0  0]
 [ 0 17  1  0]
 [ 0  4 13  0]
 [ 1  3  1  0]]

The mlp Score is  0.8088235294117647
The confusion matrix is  [[28  0  0  0]
 [ 2 15  1  0]
 [ 1  4 12  0]
 [ 1  3  1  0]]

The mlp Score is  0.80

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.8
The std of the mlp scores is = 0.05


On an average, the new models for X2 and X1 performed similarly, and better than the former models. Although, the Standard Deviations are larger than in the previous models.

##### Model 3 extra

In [ ]:
#CHANGING THE number OF HIDDEN LAYERS, using X2
# Creating the neural network
#using a for loop to run through 20 times
X2_3 = np.copy(X2)

X_train, X_test, y_train, y_test = train_test_split(X2_3, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    mlp = MLPClassifier(hidden_layer_sizes=(30,30,30,30,30), max_iter=1000)
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.8235294117647058
The confusion matrix is  [[29  0  0  0]
 [ 2 18  1  0]
 [ 1  1  8  1]
 [ 1  5  0  1]]

The mlp Score is  0.8088235294117647
The confusion matrix is  [[29  0  0  0]
 [ 4 17  0  0]
 [ 1  1  9  0]
 [ 5  1  1  0]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[29  0  0  0]
 [ 5 14  2  0]
 [ 1  1  9  0]
 [ 4  3  0  0]]

The mlp Score is  0.6911764705882353
The confusion matrix is  [[29  0  0  0]
 [ 7 10  4  0]
 [ 1  2  8  0]
 [ 4  2  1  0]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[29  0  0  0]
 [ 7 13  1  0]
 [ 1  3  7  0]
 [ 4  2  1  0]]

The mlp Score is  0.7352941176470589
The confusion matrix is  [[26  3  0  0]
 [ 1 18  2  0]
 [ 1  4  6  0]
 [ 4  2  1  0]]

The mlp Score is  0.6470588235294118
The confusion matrix is  [[21  8  0  0]
 [ 1 14  6  0]
 [ 1  1  9  0]
 [ 3  3  1  0]]

The mlp Score is  0.75
The confusion matrix is  [[25  4  0  0]
 [ 1 20  0  0]
 [ 0  5  6  0]
 [ 2  4  1  0]]

The mlp Score is  0.77

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.76
The std of the mlp scores is = 0.04


This model for X2 performed worse of all after increasing the number of hidden layers to 5.

##### Model 4 extra

In [ ]:
#CHANGING THE number and size OF HIDDEN LAYER, using X2
# Creating the neural network
#using a for loop to run through 20 times
X2_4 = np.copy(X2)

X_train, X_test, y_train, y_test = train_test_split(X2_4, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    #test something seen in a research paper. Use just 1 hidden layer
    mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000)
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.8676470588235294
The confusion matrix is  [[24  0  0  0]
 [ 3 24  1  0]
 [ 2  3  9  0]
 [ 0  0  0  2]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[24  0  0  0]
 [11 16  1  0]
 [ 2  3  9  0]
 [ 1  1  0  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[24  0  0  0]
 [10 17  1  0]
 [ 2  2 10  0]
 [ 0  0  0  2]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[24  0  0  0]
 [ 8 19  1  0]
 [ 2  3  9  0]
 [ 1  1  0  0]]

The mlp Score is  0.6911764705882353
The confusion matrix is  [[24  0  0  0]
 [12 14  2  0]
 [ 3  2  9  0]
 [ 1  1  0  0]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[24  0  0  0]
 [ 5 22  1  0]
 [ 2  4  8  0]
 [ 1  1  0  0]]

The mlp Score is  0.8088235294117647
The confusion matrix is  [[24  0  0  0]
 [ 6 20  2  0]
 [ 2  3  9  0]
 [ 0  0  0  2]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[24  0  0  0]
 [10 17  1  0]
 [ 2  4  8  0]
 [ 1  1  0  0]]

The mlp 

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.78
The std of the mlp scores is = 0.05


Wow! The assertion in the research paper did not do too bad as this X2 model did better compared with the other models. Although the Standard deviation is high in comparison.
This model performed better than when the number of hidden layers was increased above!

##### Model 5 extra

In [ ]:
#CHANGING THE number and size OF HIDDEN LAYER, using X2
# Creating the neural network
#using a for loop to run through 20 times
X2_5 = np.copy(X2)

X_train, X_test, y_train, y_test = train_test_split(X2_5, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    #test something seen in a research paper. Use just 1 hidden layer
    mlp = MLPClassifier(hidden_layer_sizes=(11,), max_iter=10000)#chose 11 by taking the mean of the number input features and output classes. 
    #Use 10,000 due to convergence warning for 1000
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.7794117647058824
The confusion matrix is  [[28  0  0  0]
 [ 3 14  0  0]
 [ 2  3 11  0]
 [ 1  4  2  0]]

The mlp Score is  0.7352941176470589
The confusion matrix is  [[28  0  0  0]
 [ 4 13  0  0]
 [ 1  6  9  0]
 [ 1  3  3  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[28  0  0  0]
 [ 3 13  1  0]
 [ 1  3 12  0]
 [ 1  3  3  0]]

The mlp Score is  0.7352941176470589
The confusion matrix is  [[28  0  0  0]
 [ 4 12  1  0]
 [ 1  5 10  0]
 [ 1  3  3  0]]

The mlp Score is  0.7058823529411765
The confusion matrix is  [[28  0  0  0]
 [ 7  9  1  0]
 [ 1  4 11  0]
 [ 3  2  2  0]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[28  0  0  0]
 [ 5  9  2  1]
 [ 1  3 12  0]
 [ 2  2  3  0]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[28  0  0  0]
 [ 5 11  1  0]
 [ 1  2 13  0]
 [ 1  3  3  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[28  0  0  0]
 [ 2 14  1  0]
 [ 1  4 11  0]
 [ 1  3  3  0]]

The mlp 

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.75
The std of the mlp scores is = 0.04


This model performs very similarly to the other models, and with the most common standard deviation of 0.04. Using the default 100 hidden layers performed better than using the average of the input and output features.

###### Model 6 extra

In [ ]:
#reduce the features in X2 to just 12

In [ ]:
#create new array
new_X2 = [] # inputs
y = [] # true values

# Here, we iterate over the lines of the file. If a line matches one of our classes, we split the next line 

for i in range(len(raw_data) - 1):
    data2 = raw_data[i].strip() # .strip() removes the line endings 
    if data2 in classes.keys(): # If the line in the data matches one of our classes (for eg, 'normal')
        features = [int(x) for x in raw_data[i+1].strip().split('\t') + raw_data[i+5].strip().split('\t')] # Split the next line to get our features
        new_X2.append(features)
        y.append(classes[data2]) # And record which class this set of features belongs to

In [ ]:
new_X2[0]

[-1, -1, 63, -3, -1, 0, -1, -1, 63, -3, -1, 0]

In [ ]:
#CHANGING THE number and size OF HIDDEN LAYER, using a new data set according to X2
# Creating the neural network
#using a for loop to run through 20 times
new_X2_1 = np.copy(new_X2)

X_train, X_test, y_train, y_test = train_test_split(new_X2_1, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    #use a size of 100 with one hidden layer as this performed good so far
    mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000)#chose 11 by taking the mean of the number input features and output classes. 
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.8088235294117647
The confusion matrix is  [[29  0  0  0]
 [ 3 17  3  0]
 [ 1  0  9  0]
 [ 3  1  2  0]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[29  0  0  0]
 [ 2 18  3  0]
 [ 2  1  7  0]
 [ 2  3  1  0]]

The mlp Score is  0.8235294117647058
The confusion matrix is  [[29  0  0  0]
 [ 1 19  3  0]
 [ 1  1  8  0]
 [ 2  2  2  0]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[29  0  0  0]
 [ 9 12  2  0]
 [ 2  1  7  0]
 [ 3  1  1  1]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[29  0  0  0]
 [ 6 13  4  0]
 [ 1  1  8  0]
 [ 2  1  1  2]]

The mlp Score is  0.6764705882352942
The confusion matrix is  [[29  0  0  0]
 [11 10  2  0]
 [ 3  0  7  0]
 [ 3  1  2  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[29  0  0  0]
 [ 3 18  2  0]
 [ 1  3  6  0]
 [ 2  2  2  0]]

The mlp Score is  0.7647058823529411
The confusion matrix is  [[29  0  0  0]
 [ 3 17  3  0]
 [ 2  2  6  0]
 [ 1  3  2  0]]

The mlp 

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.76
The std of the mlp scores is = 0.04


This model with less features performs better than that of 18 features, despite using the same hidden_layer parameters and number of iterations, and with the lowest standard deviation so far of 0.04. 

###### Model 7 extra

In [ ]:
#increase the features in X2 to just 24

In [ ]:
#create new array
newer_X2 = [] # inputs
y = [] # true values

# Here, we iterate over the lines of the file. If a line matches one of our classes, we split the next line 

for i in range(len(raw_data) - 1):
    data2 = raw_data[i].strip() # .strip() removes the line endings 
    if data2 in classes.keys(): # If the line in the data matches one of our classes (for eg, 'normal')
        features = [int(x) for x in raw_data[i+1].strip().split('\t') + 
                    raw_data[i+5].strip().split('\t') + 
                    raw_data[i+10].strip().split('\t') + 
                    raw_data[i+15].strip().split('\t')] # Split the next line to get our features
        newer_X2.append(features)
        y.append(classes[data2]) # And record which class this set of features belongs to

In [ ]:
newer_X2[0]

[-1,
 -1,
 63,
 -3,
 -1,
 0,
 -1,
 -1,
 63,
 -3,
 -1,
 0,
 -1,
 -1,
 61,
 -3,
 0,
 0,
 -1,
 0,
 64,
 -2,
 -1,
 0]

In [ ]:
#CHANGING THE number and size OF HIDDEN LAYER, using another data set created according to X2
# Creating the neural network
#using a for loop to run through 20 times
newer_X2_1 = np.copy(newer_X2)

X_train, X_test, y_train, y_test = train_test_split(newer_X2_1, y)

# Prepare to scale the inputs
scaler = StandardScaler()
scaler.fit(X_train)

# Scaling the inputs 
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)



mlpscores = [] #store scores here

for counter in range(1,21): #repeat process in loop 20 times
    #use a size of 100 with one hidden layer as this performed good so far
    mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000)#chose 11 by taking the mean of the number input features and output classes. 
    mlp.fit(X_train,y_train)
    predictions = mlp.predict(X_test)
    
    scores = mlp.score(X_test, y_test)
    mlpscores += [scores]
    
    print('The mlp Score is ',scores)
    # The confusion matrix gives a better picture of what errors are slipping in
    print('The confusion matrix is ',confusion_matrix(y_test,predictions))
    print("")

C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\acer\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int32 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


The mlp Score is  0.7647058823529411
The confusion matrix is  [[23  0  0  0]
 [ 5 17  0  0]
 [ 0  5 12  0]
 [ 1  3  2  0]]

The mlp Score is  0.75
The confusion matrix is  [[23  0  0  0]
 [ 6 16  0  0]
 [ 1  4 11  1]
 [ 1  3  1  1]]

The mlp Score is  0.7205882352941176
The confusion matrix is  [[23  0  0  0]
 [ 7 15  0  0]
 [ 1  5 11  0]
 [ 1  3  2  0]]

The mlp Score is  0.7794117647058824
The confusion matrix is  [[23  0  0  0]
 [ 7 15  0  0]
 [ 1  3 13  0]
 [ 2  1  1  2]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[23  0  0  0]
 [ 6 16  0  0]
 [ 1  3 13  0]
 [ 1  2  1  2]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[23  0  0  0]
 [ 5 17  0  0]
 [ 0  4 13  0]
 [ 0  4  1  1]]

The mlp Score is  0.8088235294117647
The confusion matrix is  [[23  0  0  0]
 [ 5 17  0  0]
 [ 0  4 13  0]
 [ 1  2  1  2]]

The mlp Score is  0.7941176470588235
The confusion matrix is  [[22  1  0  0]
 [ 4 18  0  0]
 [ 0  4 13  0]
 [ 0  4  1  1]]

The mlp Score is  0.75

In [ ]:
#compute average of mlp scores obtained
average = np.mean(mlpscores) #calculate average
print ("The average mlp score is =", round(average, 2)) 

#compute std of mlp scores obtained
std = np.std(mlpscores)
print ("The std of the mlp scores is =", round(std, 2)) 

The average mlp score is = 0.79
The std of the mlp scores is = 0.03


The average mlp score is much more better with 24 features, and also has a lower Standard Deviation of 0.03!

##### The 7th model above is the best model with an average mlp score of 0.79, and an STD (over 20 repeated runs) of 0.03

End